### ISSUE TITLE
improve selection of matching overload for annotated-self inference

### ISSUE THREAD

**Issue by posita** [astral-sh/ty#3330]
### Summary

When a class (e.g., `Box` in our example below) has two `__init__` overloads — one taking `Iterable[Never]` (binding `self` to `Box[Never]`) and one taking `Iterable[_T]` (binding `self` to `Box[_T]`) — passing a value whose iteration type is `Any` incorrectly resolves to the `Iterable[Never]` overload, yielding `Box[Never]` instead of `Box[_T]`.

[Playground](https://play.ty.dev/581c0167-696a-4f6f-8c3e-7fcf4ae4304f)

```python
# any_never_bug.py
from enum import IntEnum
from typing import Generic, Iterable, Never, TypeVar, overload, reveal_type

_T = TypeVar("_T")


class Box(Generic[_T]):
    @overload
    def __init__(self: "Box[Never]", val: Iterable[Never]) -> None: ...
    @overload
    def __init__(self: "Box[_T]", val: Iterable[_T]) -> None: ...
    def __init__(self, val: object) -> None: ...


class Color(IntEnum):
    RED = 1
    GREEN = 2


reveal_type(Box(Color))  # Expected: Box[Color], mypy gives: Box[Never]
```

```sh
% ty check --config-file /dev/null any_never_bug.py
info[revealed-type]: Revealed type
  --> any_never_bug.py:21:13
   |
21 | reveal_type(Box(Color))  # Expected: Box[Color], mypy gives: Box[Never]
   |             ^^^^^^^^^^ `Box[Never]`
   |

Found 1 diagnostic
```

Expected: `Box[Color]`
Actual: `Box[Never]`

Notes:

My best guess on root cause is that `EnumMeta.__iter__` is typed in typeshed as returning `Iterator[Any]`, so `Color` (the class) is seen as `Iterable[Any]`. I'm guessing that since `Any` is consistent with `Never`, mypy selects the `Iterable[Never]` overload rather than the more general `Iterable[_T]` one. However, `Iterable[Any]` should probably not be considered assignable to `Iterable[Never]`. I'm pretty sure `Never` is the bottom type and nothing except `Never` itself should be a subtype of it. Pyright and Pyrefly both correctly infer `Box[Color]` for this case.

Ref: python/mypy#21317

### Version

ty 0.0.32

**Comment by carljm** [comment]
I filed https://github.com/astral-sh/ty/issues/3331 for the issue of how we handle an enum type assigned to `Iterable[T]`. I think that fixing that will be sufficient to get the result you want on the overall example? 

Here I'd like to focus on the overload handling. So using this modification of the above example, to remove the enum factor and directly use an argument of type `Iterable[Any]`:

```py
from typing import Generic, Iterable, Never, TypeVar, overload, reveal_type, Any

_T = TypeVar("_T")

class Box(Generic[_T]):
    @overload
    def __init__(self: "Box[Never]", val: Iterable[Never]) -> None: ...
    @overload
    def __init__(self: "Box[_T]", val: Iterable[_T]) -> None: ...
    def __init__(self, val: object) -> None: ...

def _(i: Iterable[Any]):
    reveal_type(Box(i))
```

I think the use of typevars in this example is actually incorrect, and pyright/pyrefly both complain about it. You are not allowed to use a class-scoped type variable in a `self` annotation. I think this should be expressed with a separate type variable scoped to the `__init__` overload:

```py
from typing import Generic, Iterable, Never, TypeVar, overload, reveal_type, Any

_S = TypeVar("_S")
_T = TypeVar("_T")

class Box(Generic[_T]):
    @overload
    def __init__(self: "Box[Never]", val: Iterable[Never]) -> None: ...
    @overload
    def __init__(self: "Box[_S]", val: Iterable[_S]) -> None: ...
    def __init__(self, val: object) -> None: ...

def _(i: Iterable[Any]):
    reveal_type(Box(i))
```

It seems like in this example, mypy reveals `Box[Any]`, and all other type checkers currently reveal `Box[Never]`. I think it's actually not at all clear what the correct answer is here. We (and most type checkers) are just picking the `self` overload type from the first matching overload in this case. It looks to me like mypy is not taking `Any` directly from the argument, because if I change `Box[_S]` to `Box[tuple[_S]]` it still infers `Box[Any]`, not `Box[tuple[Any]]`. This suggests it's doing a sort of "gradual join" that it does for gradual ambiguous overload return types -- it's combining `Box[Never]` and `Box[whatever]` into the gradual type that can materialize to either of them, `Box[Any]`. That does seem to me like the best approach in this case.

What are you intending to achieve with this weird `Never` overload? It seems unnecessary; this version would have the desired result on all type checkers:

```py
from typing import Generic, Iterable, Never, TypeVar, reveal_type, Any

_T = TypeVar("_T")

class Box(Generic[_T]):
    def __init__(self, val: Iterable[_T]) -> None: ...

def _(i: Iterable[Any], n: Iterable[Never]):
    reveal_type(Box(i))
    reveal_type(Box(n))
```

**Comment by posita** [comment]
> I think the use of typevars in this example is actually incorrect, and pyright/pyrefly both complain about it. You are not allowed to use a class-scoped type variable in a `self` annotation.

I was under the impression that a `TypeVar` used in a `self` annotation was treated semantically differently than the same facially similar `TypeVar` used as a class-scoped one (meaning they might as well be completely different `TypeVar`s, depending on context). That assumption was likely formed after witnessing various type-checking behaviors, but I can't point to anything specific. What I should have done was created something similar to your `_S` and `_T` example originally to make things clearer.

> What are you intending to achieve with this weird `Never` overload? It seems unnecessary; this version would have the desired result on all type checkers ... .

I may have created too simple of an example. The idea is something more like this:

```python
from typing import Any, Generic, Iterable, Never, SupportsInt, TypeVar, overload, reveal_type

_SupportsIntT = TypeVar(name="_SupportsIntT", bound=SupportsInt)
_ClassT = TypeVar(name="_ClassT")

class Box(Generic[_ClassT]):
    @overload
    def __init__(self: Box[_SupportsIntT], val: Iterable[_SupportsIntT]) -> None: ...
    @overload
    def __init__(self: Box[Never], val: Iterable[Any]) -> None: ...

reveal_type(Box([0]))  # should be Box[int]
reveal_type(Box([0.0]))  # should be Box[float]
reveal_type(Box(["nope"]))  # should be Box[Never]
```

[ty playground](https://play.ty.dev/2bc5945b-0ff1-4683-9912-0fc28c156efb)
[pyright playground](https://pyright-play.net/?code=GYJw9gtgBALgngBwJYDsDmUkQWEMoCCKcANFAOICmKlISAxmQJIy0CGARgDaVkBylAG60yAZQCuCHHgDOTFDDIAVRJQBqbEGTDCQXMGwAmZEEMpsuAfXgJKAKDuWJU3DDkKlUALxQVtjSAAFChsEJReAEROktJu8jBKEWQcYOIohl7Ose4wAJSOAMJcbDIynj5%2B6prBoeFRRSVlEfl29MWlUABCYAAegVQ0dPQA2pYNpUoAurkAXHZQC1AAAjq0%2Bkbzi4aUwFCWlqhIMPuBMpRcwDNdvaNZrjlTZIIWVyzs3JS3MffxU7lQAFoAHxQPhgGhXAB00M2CxWunWhlhUG2u32h2OllO50u1x6wwEukmTxeUDeIE4PGGRDg00BILBEKg0MhDlMwgs1lUgW6fWGAAZpv8oABiKAyAAWqS4higHEoeOGqBgkzs7PMVhslB5vUCAshgtywrFkulsvliuA6xVarMnK1Or5ERQYFsESFCxNUvEMrlCt5BLMIEmQA)

**Comment by carljm** [comment]
So instead of a type error on `Box(["nope"])` (from `"nope"` not being a valid `_SupportsIntT`), you instead want to silence the type error and just infer `Box[Never]` instead?

**Comment by posita** [comment]
> So instead of a type error on `Box(["nope"])` (from `"nope"` not being a valid `_SupportsIntT`), you instead want to silence the type error and just infer `Box[Never]` instead?

Actually, no. I am confusing two separate situations here. I think #3331 captures what is needed. 👍 Closing anything that remains here as either already addressed or a non sequitur at this point.
